core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -821,6 +821,12 @@ public ConnectivityState getState(boolean requestConnection) {
             @Override
             public void run() {
               exitIdleMode();
+              if (subchannelPicker != null) {
+                if (shutdown.get() || panicMode) {
```

Comment:
This looks unnecessary. I would really hope this is unnecessary. We don't want to be checking these all over the place.

Reply:
We can probably instead check channelStateManager.getState() again here.

Reply:
If `shutdownNameResolverAndLoadBalancer()` has been called, which is done by both shutdown and panic, then `subchannelPicker` will have been cleared. If in `panicMode`, it will point to a `panicPickResult` which should do nothing. There's no risk in using the `subchannelPicker` incorrectly now, and that should be maintained in the future.

Code _setting_ subchannelPicker could reasonably need to look at those fields. But readers shouldn't need to, as a rule.

Reply:
`shutdownNameResolverAndLoadBalancer()` is called when the channel is terminating, not immediately after `shutdown()`. That said, even when state == SHUTDOWN and `subchannelPicker` is not null, there is no harm calling `requestConnection()`. Yes I agree we can remove this `if`.